### PR TITLE
shading and overlay

### PR DESCRIPTION
**Pull request by e2002e**
Overlay takes shading into account, so the emission is working as the issue #2694 mentioned.
screen
screen

### CODE REVIEW

**Comment by e2002e** [comment]
Maybe we should add the overlay pass before the compositor to include the meshes in compositor pass ? If there is no reason to not do it, tell me.

**Comment by MoritzBrueckner** [comment]
Hi, it's a good idea to also implement shading for overlay meshes, I probably just would have added the emission color and kept the shadeless output for now :D

However, I'm not sure whether we should re-run the entire deferred render step here, perhaps it makes more sense to render overlay meshes in a forward fashion. Assuming that overlay meshes probably take only a small part of the screen, it might waste too much GPU bandwidth to use the gbuffer, but this is something that needs to be profiled.

As for the reason that overlay meshes are rendered after the compositor pass (and other post processing effects) I would assume it's because of their potential use in UIs, but then there should still be some kind of tone mapping for example (and sometimes one maybe wants to include them in post processing?). Maybe we need to split the compositor pass up... @luboslenco Do you know/remember the purpose of the overlay option? Also, is it by purpose that overlay meshes were shadeless before this PR, for UI e.g.?
It makes sense of course to draw the overlay meshes after transparent meshes (and thus after deferred lighting) in order to really overlay them.

Did you test this on the forward RP? The code looks like it only works on deferred and might even break on forward rendering.

File: blender/arm/material/make_shader.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def build(material: Material, mat_users: Dict[Material, List[Object]], mat_armus
             con = make_transluc.make(rp)
 
         elif rp == 'overlay':
-            con = make_overlay.make(rp)
+            con = make_mesh.make(rp, rpasses)
```
Comment (MoritzBrueckner) [inline blender/arm/material/make_shader.py:100]: Why did you change this here? Also, isn't make_overlay.py completely obsolete then since it's no longer called? What's the reason for the change in make_overlay.py then?

File: blender/arm/material/make_overlay.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def make(context_id):
     arm_discard = mat.arm_discard
     is_transluc = mat_utils.is_transluc(mat)
     parse_opacity = (blend and is_transluc) or arm_discard
-    make_mesh.make_base(con_overlay, parse_opacity=parse_opacity)
+    make_mesh.make_deferred(con_overlay, parse_opacity=parse_opacity)
```
Comment (MoritzBrueckner) [inline blender/arm/material/make_overlay.py:33]: I don't think this will work on forward RP.

File: Sources/armory/renderpath/RenderPathDeferred.hx [hunk 3]
```diff
@@ -626,6 +646,12 @@ class RenderPathDeferred {
 		}
 		#end
 
+		#if ((rp_voxelgi_shadows) || (rp_voxelgi_refraction))
+		{
+			path.bindTarget(voxels, "voxels");
+		}
+		#end
+
```
Comment (MoritzBrueckner) [inline Sources/armory/renderpath/RenderPathDeferred.hx:654]: Are these changes really part of the change of this PR or are they accidentally committed because of your Voxel GI PR?

File: Sources/armory/renderpath/RenderPathDeferred.hx [hunk 4]
```diff
@@ -53,7 +52,14 @@ class RenderPathDeferred {
...
 		{
 			Inc.initGI();
+			#if arm_voxelgi_temporal
+			{
+				Inc.initGI("voxelsB");
+			}
+			#end
+			#if (rp_gi == "Voxel AO")
 			path.loadShader("shader_datas/deferred_light/deferred_light_VoxelAOvar");
+			#end
```
Comment (MoritzBrueckner) [inline Sources/armory/renderpath/RenderPathDeferred.hx:62]: Same question here, will this even work without your Voxel GI PR?

File: Sources/armory/renderpath/RenderPathDeferred.hx (hunk 3 above)
Comment (MoritzBrueckner) [reply Sources/armory/renderpath/RenderPathDeferred.hx:654]: Looks like GitHub didn't like my rather large selection range, this not only refers to the 4 lines that are shown but also to the voxel-related lines above that.

File: blender/arm/material/make_shader.py (hunk 1 above)
Comment (e2002e) [reply blender/arm/material/make_shader.py:100]: shader needs to fill datas for the deferred shader

File: blender/arm/material/make_overlay.py (hunk 2 above)
Comment (e2002e) [reply blender/arm/material/make_overlay.py:33]: Ok I'll find a workaround

File: Sources/armory/renderpath/RenderPathDeferred.hx (hunk 3 above)
Comment (e2002e) [reply Sources/armory/renderpath/RenderPathDeferred.hx:654]: No only AO

File: Sources/armory/renderpath/RenderPathDeferred.hx (hunk 4 above)
Comment (e2002e) [reply Sources/armory/renderpath/RenderPathDeferred.hx:62]: yes

**Comment by e2002e** [comment]
> I'm not sure whether we should re-run the entire deferred render step here

The deferred shader, If I'm not mistaken, will only render on the bind texture, which is set to buf, and only contains the newly drawn overlay meshes.

**Comment by MoritzBrueckner** [comment]
Sorry, I think I was a bit imprecise here, I didn't mean that it re-renders the entire scene again, but that you do another deferred lighting pass. I fear that this could be detrimental for performance because overlay meshes are likely to only fill small parts of the screen most of the time. There would be many unused fragments in the buffers that maybe waste memory or cause cache misses. I'm not a GPU expert so I might be wrong here, but I think this is something you should keep in mind (for translucent meshes we always use forward rendering for example, but this is obviously a slightly different case).

**Comment by e2002e** [comment]
No, I was incorrect when saying that the target is what is rendered by the shader. I know the deferred shader deals with gbuffer0 and gbuffer1. These are overridden, when target is set to overlay. Kein Problems.

**Comment by e2002e** [comment]
The issue you talk about is that when drawing overlay meshes via deferred shader, the rest of the scene can be present in the buffers. It is not because we clear the buffers beforehand.
I took into account the fact that overlay can serve as UI elements, so I'm going to add an option to shade these or not.
